File: Algorithms/src/2416. Sum of Prefix Scores of Strings/Code/solution.java
import java.util.HashMap;

// Class representing a single node in the Trie
class TrieNode {
    // HashMap to store child nodes for each character
    HashMap<Character, TrieNode> children = new HashMap<>();

    // Count stores how many times this node has been visited (useful for scoring
    // prefixes)
    int count = 0;
}

// Class representing the Trie (prefix tree) structure
class Trie {
    TrieNode root; // Root node of the Trie

    // Constructor to initialize the Trie with an empty root node
    public Trie() {
        root = new TrieNode();
    }

    // Method to insert a word into the Trie
    public void insert(String word) {
        TrieNode node = root; // Start at the root node

        // Loop through each character in the word
        for (char ch : word.toCharArray()) {
            // If the current character isn't already a child of the current node, add it
            node.children.putIfAbsent(ch, new TrieNode());

            // Move to the child node corresponding to the current character
            node = node.children.get(ch);

            // Increment the count to signify that this node has been visited
            node.count++;
        }
    }

    // Method to get the sum of scores for all prefixes of a given word
    public int getPrefixScoreSum(String word) {
        TrieNode node = root; // Start at the root node
        int scoreSum = 0; // Initialize the sum of prefix scores

        // Loop through each character in the word
        for (char ch : word.toCharArray()) {
            // Move to the child node corresponding to the current character
            node = node.children.get(ch);

            // Add the count of the current node to the score sum
            scoreSum += node.count;
        }

        // Return the total sum of scores for all prefixes of the word
        return scoreSum;
    }
}

// Solution class that solves the problem using the Trie
class Solution {
    // Method to calculate the sum of prefix scores for each word in the array
    public int[] sumPrefixScores(String[] words) {
        Trie trie = new Trie(); // Create a new Trie

        // Insert all words from the array into the Trie
        for (String word : words) {
            trie.insert(word);
        }

        // Create an array to store the result (sum of prefix scores for each word)
        int[] result = new int[words.length];

        // Calculate the sum of prefix scores for each word
        for (int i = 0; i < words.length; i++) {
            result[i] = trie.getPrefixScoreSum(words[i]);
        }

        // Return the result array
        return result;
    }
}

File: Algorithms/src/2416. Sum of Prefix Scores of Strings/Code/solution.py
# Class representing a node in the Trie
class TrieNode:
    def __init__(self):
        # Dictionary to hold child nodes, where keys are characters
        self.children = {}
        # Count to store how many times this node (prefix) is visited
        self.count = 0

# Class representing the Trie (prefix tree) structure
class Trie:
    def __init__(self):
        # The root node of the Trie
        self.root = TrieNode()

    # Method to insert a word into the Trie
    def insert(self, word: str):
        node = self.root  # Start from the root node
        # Loop through each character of the word
        for ch in word:
            # If the character is not a child of the current node, add it
            if ch not in node.children:
                node.children[ch] = TrieNode()
            # Move to the next node (child node corresponding to the character)
            node = node.children[ch]
            # Increment the count for the current node (this prefix)
            node.count += 1

    # Method to calculate the sum of prefix scores for a given word
    def get_prefix_score_sum(self, word: str) -> int:
        node = self.root  # Start from the root node
        score_sum = 0     # Initialize score sum
        # Loop through each character in the word
        for ch in word:
            # Move to the next node (child node corresponding to the character)
            node = node.children[ch]
            # Add the count of the current node (this prefix) to the score sum
            score_sum += node.count
        return score_sum

# Class to solve the problem of finding prefix scores for a list of words
class Solution:
    def sumPrefixScores(self, words: List[str]) -> List[int]:
        trie = Trie()  # Initialize a new Trie instance
        
        # Insert all words into the Trie to build the structure
        for word in words:
            trie.insert(word)
        
        # For each word, calculate the sum of prefix scores and return the result as a list
        return [trie.get_prefix_score_sum(word) for word in words]